File: Lab_5/Rubayed/Codes/Edge_Detection.m
% Read the image
img = imread('baby2.jpg'); % Replace 'your_image.jpg' with your image file
% Check if the image is already grayscale
if size(img, 3) == 3
    grayImg = rgb2gray(img); % Convert to grayscale if it's a color image
else
    grayImg = img; % Use the image as is if it's already grayscale
end

% Prewitt Edge Detection
prewitt_x = fspecial('prewitt'); % Horizontal Prewitt filter
prewitt_y = prewitt_x'; % Vertical Prewitt filter
prewitt_edge = sqrt(imfilter(double(grayImg), prewitt_x).^2 + imfilter(double(grayImg), prewitt_y).^2);

% Sobel Edge Detection
sobel_x = fspecial('sobel'); % Horizontal Sobel filter
sobel_y = sobel_x'; % Vertical Sobel filter
sobel_edge = sqrt(imfilter(double(grayImg), sobel_x).^2 + imfilter(double(grayImg), sobel_y).^2);

% Roberts Edge Detection
roberts_x = [1 0; 0 -1]; % Roberts horizontal filter
roberts_y = [0 1; -1 0]; % Roberts vertical filter
roberts_edge = sqrt(imfilter(double(grayImg), roberts_x).^2 + imfilter(double(grayImg), roberts_y).^2);

% Isotropic Edge Detection (Laplacian of Gaussian)
h = fspecial('gaussian', [5 5], 1); % Gaussian filter
smoothedImg = imfilter(double(grayImg), h); % Smooth the image
laplacian = fspecial('laplacian', 0.5); % Laplacian filter
log_edge = imfilter(smoothedImg, laplacian);

% Display results
figure;
subplot(3,2,1); imshow(img); title('Real Image');
subplot(3,2,2); imshow(prewitt_edge, []); title('Prewitt Edge Detection');
subplot(3,2,3); imshow(sobel_edge, []); title('Sobel Edge Detection');
subplot(3,2,4); imshow(roberts_edge, []); title('Roberts Edge Detection');
subplot(3,2,5); imshow(log_edge, []); title('Laplacian of Gaussian Edge Detection');

% Adjust figure properties
sgtitle('Edge Detection Methods');

File: Lab_5/Rubayed/Codes/edge-detection.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

# Read the image
img = cv2.imread('baby2.jpg')  # Replace with your image file
# Convert to RGB (OpenCV loads images in BGR format)
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

# Check if the image is already grayscale
if len(img.shape) == 3:
    gray_img = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)  # Convert to grayscale if it's a color image
else:
    gray_img = img  # Use the image as is if it's already grayscale

# Prewitt Edge Detection
prewitt_x = np.array([[1, 0, -1], [1, 0, -1], [1, 0, -1]])
prewitt_y = np.array([[1, 1, 1], [0, 0, 0], [-1, -1, -1]])
prewitt_edge_x = cv2.filter2D(gray_img.astype(float), -1, prewitt_x)
prewitt_edge_y = cv2.filter2D(gray_img.astype(float), -1, prewitt_y)
prewitt_edge = np.sqrt(prewitt_edge_x**2 + prewitt_edge_y**2)

# Sobel Edge Detection
sobel_x = cv2.Sobel(gray_img.astype(float), cv2.CV_64F, 1, 0, ksize=5)
sobel_y = cv2.Sobel(gray_img.astype(float), cv2.CV_64F, 0, 1, ksize=5)
sobel_edge = np.sqrt(sobel_x**2 + sobel_y**2)

# Roberts Edge Detection
roberts_x = np.array([[1, 0], [0, -1]])
roberts_y = np.array([[0, 1], [-1, 0]])
roberts_edge_x = cv2.filter2D(gray_img.astype(float), -1, roberts_x)
roberts_edge_y = cv2.filter2D(gray_img.astype(float), -1, roberts_y)
roberts_edge = np.sqrt(roberts_edge_x**2 + roberts_edge_y**2)

# Isotropic Edge Detection (Laplacian of Gaussian)
gaussian_blur = cv2.GaussianBlur(gray_img.astype(float), (5, 5), 1)
log_edge = cv2.Laplacian(gaussian_blur, cv2.CV_64F)

# Display results
plt.figure(figsize=(10, 10))
plt.subplot(3, 2, 1); plt.imshow(img.astype(np.uint8)); plt.title('Real Image'); plt.axis('off')
plt.subplot(3, 2, 2); plt.imshow(prewitt_edge, cmap='gray'); plt.title('Prewitt Edge Detection'); plt.axis('off')
plt.subplot(3, 2, 3); plt.imshow(sobel_edge, cmap='gray'); plt.title('Sobel Edge Detection'); plt.axis('off')
plt.subplot(3, 2, 4); plt.imshow(roberts_edge, cmap='gray'); plt.title('Roberts Edge Detection'); plt.axis('off')
plt.subplot(3, 2, 5); plt.imshow(log_edge, cmap='gray'); plt.title('Laplacian of Gaussian Edge Detection'); plt.axis('off')

# Adjust figure properties
plt.suptitle('Edge Detection Methods', fontsize=16)
plt.tight_layout()
plt.show()
